examples/json.dx

```diff
@@ -0,0 +1,258 @@
+
+
+data JValue = AsJValue String
+-- TODO - once Dex supports recursive ADT JValue becomes Value.
+
+data Value = 
+     AsObject (List (String & JValue))
+     AsArray (List JValue)
+     AsString String
+     AsFloat Float
+     AsInt Int
+
+interface ToJSON a
+  toJSON : a -> Value
+
+
+-- These three are private methods. Users should use type classes.
+def escape (x:JValue) : String =
+    (AsJValue y) = x
+    y
+
+
+def collapse (x:Value) : JValue =
+    AsJValue $ case x of
+         AsString y ->  "\"" <> y <> "\""
+         AsFloat y -> show y
+         AsInt y -> show y
+         AsObject (AsList _ y) ->
+              ("{" <> (join ", " $ for i.
+                 (k, v) = y.i 
+                 "\"" <> k <> "\"" <> ":" <> (escape v)) <> "}")
+         AsArray (AsList _ y) -> ("[" <> (join ", " $ for i. escape y.i) <> "]")
+
+def hide [ToJSON a] (x:a) : JValue =
+    collapse $ toJSON x
+
+instance Show Value 
+  show = \x. escape $ collapse x
+
+instance ToJSON String
+  toJSON = AsString
+  
+instance ToJSON Int
+  toJSON = AsInt
+  
+instance ToJSON Float
+  toJSON = AsFloat
+  
+instance ToJSON Value
+  toJSON = id
+  
+instance [ToJSON v] ToJSON ((Fin n) => v)
+  toJSON = \x . AsArray $ AsList _ $ for i. hide x.i
+
+instance [ToJSON v] ToJSON ((Fin n) => (String & v))
+  toJSON = \x . AsObject $ AsList _ $ for i. (fst x.i, hide $ snd x.i)
+
+instance [ToJSON v] ToJSON (List (String & v))
+  toJSON = \(AsList _ x) . toJSON x
+
+
+' ## Graph Grammars (Vega-Lite Spec)
+
+Options = List (String & String)
+
+data Mark =
+     AsMark String Options
+
+data EncodingType =
+     Quantitative
+     Nominal
+     Ordinal
+
+instance Show EncodingType
+         show = (\ x.
+              case x of
+                   Quantitative -> "quantitative"
+                   Nominal -> "nominal"
+                   Ordinal -> "ordinal")
+                   
+data Channel =
+     Y
+     X
+     Color
+     Tooltip
+     HREF
+     Row
+     Col
+     Size
+     
+instance Show Channel
+         show = (\ x.
+              case x of
+                   Y -> "y"
+                   X -> "x"
+                   Color -> "color"
+                   Tooltip -> "tooltip"
+                   HREF -> "href"
+                   Size -> "size"
+                   Row -> "row"
+                   Col -> "col")
+
+data Encoding key =
+     AsEncoding Channel key EncodingType Options
+
+def enc (c:Channel) (k:Int) (et: EncodingType) : Encoding key =
+    AsEncoding c (k@_) et mempty
+
+def mark (m:String) : Mark =
+    AsMark m mempty
+
+def optsList (x:Options) : List (String & Value) =
+    (AsList _  tab) = x
+    AsList _ $ for i. (fst tab.i, toJSON $ snd tab.i)
+
+
+data DataFrame key n value =
+     AsDataFrame (n => key => value) (key => String)
+
+
+def chart [ToJSON v] (x: DataFrame (Fin key) n v)
+                     (mark: Mark)
+                     (encs : (Fin m) => Encoding (Fin key))
+                     (opts : Options)
```

Comment:
I'm having some trouble reviewing this because I don't know what `mark` and `opts` are.

Reply:
Removed